// ngx-extended-pdf-viewer/esm5/lib/events/find-result.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/events/find-result.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var FindState = {
    FOUND: 0,
    NOT_FOUND: 1,
    WRAPPED: 2,
    PENDING: 3,
};
export { FindState };
FindState[FindState.FOUND] = 'FOUND';
FindState[FindState.NOT_FOUND] = 'NOT_FOUND';
FindState[FindState.WRAPPED] = 'WRAPPED';
FindState[FindState.PENDING] = 'PENDING';
/**
 * @record
 */
export function FindResultMatchesCount() { }
if (false) {
    /** @type {?|undefined} */
    FindResultMatchesCount.prototype.current;
    /** @type {?|undefined} */
    FindResultMatchesCount.prototype.total;
    /** @type {?|undefined} */
    FindResultMatchesCount.prototype.matches;
    /** @type {?|undefined} */
    FindResultMatchesCount.prototype.matchesLength;
    /** @type {?|undefined} */
    FindResultMatchesCount.prototype.matchesColor;
}
/**
 * @record
 */
export function FindResult() { }
if (false) {
    /** @type {?} */
    FindResult.prototype.state;
    /** @type {?} */
    FindResult.prototype.previous;
    /** @type {?} */
    FindResult.prototype.matchesCount;
}
