// ngx-extended-pdf-viewer/esm5/lib/options/pdf-viewer.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/options/pdf-viewer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var ScrollModeType = {
    vertical: 0,
    horizontal: 1,
    wrapped: 2,
};
export { ScrollModeType };
ScrollModeType[ScrollModeType.vertical] = 'vertical';
ScrollModeType[ScrollModeType.horizontal] = 'horizontal';
ScrollModeType[ScrollModeType.wrapped] = 'wrapped';
/**
 * @record
 */
export function ScrollModeChangedEvent() { }
if (false) {
    /** @type {?} */
    ScrollModeChangedEvent.prototype.mode;
}
/**
 * @record
 */
export function IPDFViewer() { }
if (false) {
    /** @type {?} */
    IPDFViewer.prototype.currentPageLabel;
    /** @type {?} */
    IPDFViewer.prototype.currentPageNumber;
    /** @type {?} */
    IPDFViewer.prototype.currentScaleValue;
    /** @type {?} */
    IPDFViewer.prototype.pagesRotation;
    /** @type {?} */
    IPDFViewer.prototype.removePageBorders;
    /** @type {?} */
    IPDFViewer.prototype.scrollMode;
    /** @type {?} */
    IPDFViewer.prototype.spreadMode;
    /** @type {?} */
    IPDFViewer.prototype._pages;
}
