// ngx-extended-pdf-viewer/esm2015/lib/options/pdf-viewer-application.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/options/pdf-viewer-application.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IWebL10n() { }
if (false) {
    /**
     * @param {?} key
     * @param {?} args
     * @param {?} fallbackString
     * @return {?}
     */
    IWebL10n.prototype.get = function (key, args, fallbackString) { };
}
/**
 * @record
 */
export function ViewHistory() { }
if (false) {
    /**
     * @param {?} name
     * @param {?=} defaultValue
     * @return {?}
     */
    ViewHistory.prototype.get = function (name, defaultValue) { };
}
/**
 * @record
 */
export function FindController() { }
if (false) {
    /** @type {?} */
    FindController.prototype._pageMatches;
    /** @type {?} */
    FindController.prototype._pageMatchesColor;
    /** @type {?} */
    FindController.prototype._pageMatchesLength;
}
/**
 * @record
 */
export function IPDFViewerApplication() { }
if (false) {
    /** @type {?} */
    IPDFViewerApplication.prototype.appConfig;
    /** @type {?} */
    IPDFViewerApplication.prototype._boundEvents;
    /** @type {?} */
    IPDFViewerApplication.prototype.enablePrint;
    /** @type {?} */
    IPDFViewerApplication.prototype.eventBus;
    /** @type {?} */
    IPDFViewerApplication.prototype.findController;
    /** @type {?} */
    IPDFViewerApplication.prototype.isViewerEmbedded;
    /** @type {?} */
    IPDFViewerApplication.prototype.l10n;
    /** @type {?} */
    IPDFViewerApplication.prototype.onError;
    /** @type {?} */
    IPDFViewerApplication.prototype.page;
    /** @type {?} */
    IPDFViewerApplication.prototype.pagesCount;
    /** @type {?} */
    IPDFViewerApplication.prototype.pdfDocument;
    /** @type {?} */
    IPDFViewerApplication.prototype.pdfLinkService;
    /** @type {?} */
    IPDFViewerApplication.prototype.pdfSidebar;
    /** @type {?} */
    IPDFViewerApplication.prototype.pdfViewer;
    /** @type {?} */
    IPDFViewerApplication.prototype.printKeyDownListener;
    /** @type {?} */
    IPDFViewerApplication.prototype.sidebarViewOnLoad;
    /** @type {?} */
    IPDFViewerApplication.prototype.spreadModeOnLoad;
    /** @type {?} */
    IPDFViewerApplication.prototype.secondaryToolbar;
    /** @type {?} */
    IPDFViewerApplication.prototype.store;
    /** @type {?} */
    IPDFViewerApplication.prototype.toolbar;
    /** @type {?} */
    IPDFViewerApplication.prototype.viewer;
    /**
     * @return {?}
     */
    IPDFViewerApplication.prototype.cleanup = function () { };
    /**
     * @return {?}
     */
    IPDFViewerApplication.prototype._cleanup = function () { };
    /**
     * @return {?}
     */
    IPDFViewerApplication.prototype.close = function () { };
    /**
     * @param {?} source
     * @param {?=} options
     * @return {?}
     */
    IPDFViewerApplication.prototype.open = function (source, options) { };
    /**
     * @return {?}
     */
    IPDFViewerApplication.prototype.unbindEvents = function () { };
    /**
     * @return {?}
     */
    IPDFViewerApplication.prototype.unbindWindowEvents = function () { };
}
